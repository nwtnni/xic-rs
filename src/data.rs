pub mod token;
pub mod ast;
pub mod typ;
pub mod operand;
