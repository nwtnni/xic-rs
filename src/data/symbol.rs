use std::borrow::Cow;
use std::cell::RefCell;

use crate::Set;

thread_local! {
    /// Global cache of interned strings
    static INTERNER: RefCell<Interner> = RefCell::new(Interner::default());
}

/// Implements naive string internment.
///
/// Requires O(n) heap space to store unique strings, in
/// return for O(1) symbol equality checks and faster symbol hashing.
///
/// Does not garbage collect interned strings: the memory
/// is intentionally leaked for the duration of the program.
#[derive(Debug, Default)]
pub struct Interner(Set<&'static str>);

/// Represents a unique string.
///
/// Only the same `Interner` that produced a `Symbol` can be used
/// to resolve it to a string again.
#[derive(Copy, Clone, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub struct Symbol(usize);

impl Interner {
    /// Store `string` in this interner if not already cached.
    fn intern<'a, S>(&mut self, string: S) -> Symbol
    where
        S: Into<Cow<'a, str>>,
    {
        let cow = string.into();
        if let Some(index) = self.0.get_index_of(cow.as_ref()) {
            Symbol(index)
        } else {
            let owned = cow.into_owned().into_boxed_str();
            let leaked = Box::leak(owned);
            let (index, _) = self.0.insert_full(leaked);
            Symbol(index)
        }
    }

    /// Store static `string` in this interner if not already cached.
    fn intern_static(&mut self, string: &'static str) -> Symbol {
        let (index, _) = self.0.insert_full(string);
        Symbol(index)
    }

    /// Resolve `symbol` in this interner.
    /// Requires that `symbol` was produced by this interner.
    fn resolve(&self, symbol: Symbol) -> &'static str {
        self.0[symbol.0]
    }
}

/// Look up `string` in the global cache, and insert it if missing.
pub fn intern<'a, S>(string: S) -> Symbol
where
    S: Into<Cow<'a, str>>,
{
    INTERNER.with(|interner| interner.borrow_mut().intern(string))
}

/// Look up static `string` in the global cache, and insert it if missing.
pub fn intern_static(string: &'static str) -> Symbol {
    INTERNER.with(|interner| interner.borrow_mut().intern_static(string))
}

/// Resolve `symbol` to its string representation.
pub fn resolve(symbol: Symbol) -> &'static str {
    INTERNER.with(|interner| interner.borrow().resolve(symbol))
}

impl From<Symbol> for &'static str {
    fn from(symbol: Symbol) -> Self {
        resolve(symbol)
    }
}

impl std::str::FromStr for Symbol {
    type Err = ();
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(intern(s))
    }
}

impl std::fmt::Debug for Symbol {
    fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(fmt, "{:?}", resolve(*self))
    }
}

impl std::fmt::Display for Symbol {
    fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(fmt, "{}", resolve(*self))
    }
}
