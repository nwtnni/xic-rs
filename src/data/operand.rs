use std::sync::atomic::{AtomicUsize, Ordering};

use crate::data::symbol;

static LABELS: AtomicUsize = AtomicUsize::new(0);
static TEMPS: AtomicUsize = AtomicUsize::new(0);

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub enum Immediate {
    Constant(i64),
    Label(Label),
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub enum Label {
    Fixed(symbol::Symbol),
    Fresh(symbol::Symbol, usize),
}

impl Label {
    pub fn fresh(label: &'static str) -> Self {
        let index = LABELS.fetch_add(1, Ordering::SeqCst);
        let symbol = symbol::intern_static(label);
        Label::Fresh(symbol, index)
    }
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub enum Temporary {
    Register(Register),
    Argument(usize),
    Return(usize),
    Fresh(symbol::Symbol, usize),
}

impl Temporary {
    pub fn fresh(label: &'static str) -> Self {
        let index = TEMPS.fetch_add(1, Ordering::SeqCst);
        let symbol = symbol::intern_static(label);
        Temporary::Fresh(symbol, index)
    }
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub enum Register {
    Rax,
    Rbx,
    Rcx,
    Rdx,
    Rbp,
    Rsp,
    Rsi,
    Rdi,
    R8,
    R9,
    R10,
    R11,
    R12,
    R13,
    R14,
    R15,
}

pub trait Operand: Copy + Eq + std::hash::Hash + std::fmt::Debug + PartialOrd + Ord {}
impl Operand for Temporary {}
impl Operand for Register {}

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub enum Memory<T: Operand> {
    R(T),
    RO(T, i32),
}
