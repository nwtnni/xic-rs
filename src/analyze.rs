mod anticipated_expressions;
mod available_expressions;
mod call_graph;
mod conditional_constant_propagation;
mod constant_propagation;
mod copy_propagation;
mod dot;
mod earliest;
mod latest;
mod live_ranges;
mod live_variables;
mod postponable_expressions;
mod used_expressions;
mod valid_allocation;

pub use anticipated_expressions::AnticipatedExpressions;
pub use available_expressions::AvailableExpressions;
pub use call_graph::CallGraph;
pub use conditional_constant_propagation::ConditionalConstantPropagation;
pub use conditional_constant_propagation::Constant;
pub use conditional_constant_propagation::Reachable;
pub use constant_propagation::ConstantPropagation;
pub use copy_propagation::CopyPropagation;
pub use dot::display;
pub use earliest::Earliest;
pub use latest::Latest;
pub use live_ranges::LiveRanges;
pub use live_ranges::Range;
pub use live_variables::LiveVariables;
pub use postponable_expressions::PostponableExpressions;
pub use used_expressions::UsedExpressions;

use std::collections::VecDeque;

use petgraph::Direction;

use crate::cfg;
use crate::cfg::Cfg;
use crate::cfg::Function;
use crate::data::operand::Label;
use crate::util;
use crate::Map;
use crate::Set;

pub trait Analysis<T: Function>: Sized {
    const BACKWARD: bool;

    type Data: Clone + Eq;

    fn default(&self) -> Self::Data;

    fn default_with_metadata(&self, _label: &Label) -> Self::Data {
        self.default()
    }

    fn transfer(&self, statement: &T::Statement, output: &mut Self::Data);

    fn transfer_with_metadata(
        &self,
        _label: &Label,
        _index: usize,
        statement: &T::Statement,
        output: &mut Self::Data,
    ) {
        self.transfer(statement, output);
    }

    fn merge<'a, I>(&self, outputs: I, input: &mut Self::Data)
    where
        I: Iterator<Item = Option<&'a Self::Data>>,
        Self::Data: 'a;

    fn merge_with_metadata<'a, I>(&self, outputs: I, input: &mut Self::Data)
    where
        I: Iterator<Item = (&'a cfg::Edge, Option<&'a Self::Data>)>,
        Self::Data: 'a,
    {
        self.merge(outputs.map(|(_, data)| data), input)
    }
}

pub struct Solution<A: Analysis<T>, T: Function> {
    pub analysis: A,
    pub inputs: Map<Label, A::Data>,
    pub outputs: Map<Label, A::Data>,
}

pub fn analyze_default<A, T>(cfg: &Cfg<T>) -> Solution<A, T>
where
    A: Analysis<T> + Default,
    T: Function,
{
    analyze(Default::default(), cfg)
}

pub fn analyze<A: Analysis<T>, T: Function>(analysis: A, cfg: &Cfg<T>) -> Solution<A, T> {
    log::info!(
        "[{}] Running {} analysis on {}...",
        std::any::type_name::<Cfg<T>>(),
        std::any::type_name::<A>(),
        cfg.name(),
    );
    util::time!(
        "[{}] Done running {} analysis on {}",
        std::any::type_name::<Cfg<T>>(),
        std::any::type_name::<A>(),
        cfg.name(),
    );

    let strongly_connected_components = cfg.strongly_connected_components(A::BACKWARD);
    let (predecessors, successors) = match A::BACKWARD {
        true => (Direction::Outgoing, Direction::Incoming),
        false => (Direction::Incoming, Direction::Outgoing),
    };

    let mut inputs = Map::<Label, A::Data>::default();
    let mut outputs = Map::<Label, A::Data>::default();
    let mut worklist = VecDeque::new();
    let mut workset = Set::default();

    // Topological order
    for component in strongly_connected_components.into_iter().rev() {
        // Reverse postorder
        worklist.extend(component.iter().copied().rev());
        workset.extend(component.iter().copied().rev());

        let component = component.into_iter().collect::<Set<_>>();

        while let Some(label) = worklist.pop_front() {
            workset.remove(&label);

            let input = inputs
                .entry(label)
                .or_insert_with(|| analysis.default_with_metadata(&label));

            analysis.merge_with_metadata(
                cfg.edges_directed(predecessors, &label)
                    .map(|(predecessor, edge)| (edge, outputs.get(&predecessor))),
                input,
            );

            let mut output = input.clone();

            if A::BACKWARD {
                for (index, statement) in cfg[&label].iter().enumerate().rev() {
                    analysis.transfer_with_metadata(&label, index, statement, &mut output);
                }
            } else {
                for (index, statement) in cfg[&label].iter().enumerate() {
                    analysis.transfer_with_metadata(&label, index, statement, &mut output);
                }
            }

            match outputs.get_mut(&label) {
                Some(existing) if *existing == output => continue,
                Some(existing) => *existing = output,
                None => {
                    outputs.insert(label, output);
                }
            };

            cfg.neighbors(successors, &label)
                .filter(|successor| component.contains(successor))
                .filter(|successor| workset.insert(*successor))
                .for_each(|successor| worklist.push_back(successor));
        }
    }

    Solution {
        analysis,
        inputs,
        outputs,
    }
}
