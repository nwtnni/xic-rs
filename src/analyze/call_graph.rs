use petgraph::graphmap::DiGraphMap;
use petgraph::visit::Walker as _;
use petgraph::Direction;

use crate::data::lir;
use crate::data::operand::{Immediate, Label};
use crate::data::symbol::Symbol;

pub struct CallGraph(DiGraphMap<Symbol, ()>);

impl CallGraph {
    pub fn new<T: lir::Target>(lir: &lir::Unit<T>) -> Self {
        let mut graph = lir
            .functions
            .iter()
            .flat_map(|(caller, function)| {
                function
                    .statements
                    .iter()
                    .map(move |statement| (caller, statement))
            })
            .filter_map(|(caller, statement)| match statement {
                lir::Statement::Call(
                    lir::Expression::Immediate(Immediate::Label(Label::Fixed(callee))),
                    _,
                    _,
                ) => Some((*caller, *callee)),
                _ => None,
            })
            .collect::<DiGraphMap<_, _>>();

        for name in lir.functions.keys() {
            graph.add_node(*name);
        }

        Self(graph)
    }

    pub fn is_recursive(&self, name: &Symbol) -> bool {
        self.0.contains_edge(*name, *name)
    }

    pub fn is_leaf(&self, name: &Symbol) -> bool {
        self.0
            .neighbors_directed(*name, Direction::Outgoing)
            .count()
            == 0
    }

    pub fn postorder(&self, start: &Symbol) -> impl Iterator<Item = Symbol> + '_ {
        petgraph::visit::DfsPostOrder::new(&self.0, *start).iter(&self.0)
    }
}
