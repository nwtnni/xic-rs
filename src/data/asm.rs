use crate::data::ir;
use crate::data::operand;

pub type Unit<T> = ir::Unit<Function<T>>;
pub type Function<T> = ir::Function<Vec<Assembly<T>>>;

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Assembly<T> {
    Binary(Binary, operand::Binary<T>),
    Unary(Unary, operand::Unary<T>),
    Nullary(Nullary),
    Label(operand::Label),
    Directive(Directive),
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Directive {
    Intel,
    Align(usize),
    Local(operand::Label),
    Global(operand::Label),
    Quad(Vec<i64>),
    Data,
    Text,
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Binary {
    Add,
    Sub,
    And,
    Or,
    Xor,
    Cmp,
    Mov,
    Lea,
}

impl From<ir::Binary> for Binary {
    fn from(binary: ir::Binary) -> Self {
        match binary {
            ir::Binary::Add => Binary::Add,
            ir::Binary::Sub => Binary::Sub,
            ir::Binary::Xor => Binary::Xor,
            ir::Binary::And => Binary::And,
            ir::Binary::Or => Binary::Or,
            _ => panic!("[INTERNAL ERROR]: converting unsupported IR operator"),
        }
    }
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Unary {
    Neg,
    Push,
    Pop,
    Call { arguments: usize, returns: usize },
    Mul,
    Div(Division),
    Jmp,
    Jcc(Condition),
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Nullary {
    Cqo,
    Ret,
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Condition {
    L,
    Le,
    Ge,
    G,
    Ne,
    E,
}

impl From<ir::Condition> for Condition {
    fn from(condition: ir::Condition) -> Self {
        match condition {
            ir::Condition::Lt => Condition::L,
            ir::Condition::Le => Condition::Le,
            ir::Condition::Ge => Condition::Ge,
            ir::Condition::Gt => Condition::G,
            ir::Condition::Ne => Condition::Ne,
            ir::Condition::Eq => Condition::E,
        }
    }
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Division {
    Quotient,
    Remainder,
}
