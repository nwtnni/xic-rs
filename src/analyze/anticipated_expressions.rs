use std::collections::BTreeSet;

use crate::analyze::Analysis;
use crate::data::lir;

pub struct AnticipatedExpressions;

impl<T: lir::Target> Analysis<lir::Function<T>> for AnticipatedExpressions {
    const BACKWARD: bool = true;

    type Data = BTreeSet<lir::Expression>;

    fn new() -> Self {
        Self
    }

    fn default(&self) -> Self::Data {
        BTreeSet::new()
    }

    fn transfer(&self, statement: &lir::Statement<T>, output: &mut Self::Data) {
        match statement {
            lir::Statement::Jump(_) | lir::Statement::Label(_) => (),
            lir::Statement::CJump {
                condition: _,
                left,
                right,
                r#true: _,
                r#false: _,
            } => {
                output.insert(left.clone());
                output.insert(right.clone());
            }
            lir::Statement::Call(function, arguments, returns) => {
                for r#return in 0..*returns {
                    Self::remove(output, &lir::Expression::Return(r#return));
                }

                output.extend(arguments.iter().cloned());
                output.insert(function.clone());
            }
            lir::Statement::Move {
                destination,
                source,
            } => {
                Self::remove(output, destination);
                output.insert(source.clone());
            }
            lir::Statement::Return(returns) => {
                output.extend(returns.iter().cloned());
            }
        }
    }

    fn merge<'a, I>(&self, mut outputs: I, input: &mut Self::Data)
    where
        I: Iterator<Item = Option<&'a Self::Data>>,
        Self::Data: 'a,
    {
        input.clear();
        input.extend(outputs.next().into_iter().flatten().flatten().cloned());

        let mut stack = Vec::new();

        for output in outputs.flatten() {
            stack.extend(
                input
                    .iter()
                    .filter(|expression| !output.contains(expression))
                    // Could be avoided with https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.drain_filter
                    .cloned(),
            );

            stack.drain(..).for_each(|kill| Self::remove(input, &kill));
        }
    }
}

impl AnticipatedExpressions {
    pub(super) fn remove(output: &mut BTreeSet<lir::Expression>, kill: &lir::Expression) {
        output.remove(kill);

        let mut stack = vec![kill.clone()];

        while let Some(killed) = stack.pop() {
            output.retain(|kill| {
                if !Self::contains(kill, &killed) {
                    return true;
                }

                stack.push(kill.clone());
                false
            })
        }
    }

    fn contains(expression: &lir::Expression, killed: &lir::Expression) -> bool {
        match expression {
            lir::Expression::Argument(_)
            | lir::Expression::Return(_)
            | lir::Expression::Immediate(_)
            | lir::Expression::Temporary(_) => expression == killed,
            lir::Expression::Memory(address) => Self::contains(&*address, killed),
            lir::Expression::Binary(_, left, right) => {
                Self::contains(&*left, killed) || Self::contains(&*right, killed)
            }
        }
    }
}
