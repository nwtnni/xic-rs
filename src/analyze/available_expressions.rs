use std::collections::BTreeSet;

use crate::analyze::analyze;
use crate::analyze::Analysis;
use crate::analyze::AnticipatedExpressions;
use crate::analyze::Solution;
use crate::data::lir;
use crate::data::operand::Label;

pub struct AvailableExpressions<T: lir::Target> {
    anticipated: Solution<AnticipatedExpressions, lir::Function<T>>,
}

impl<T: lir::Target> Analysis<lir::Function<T>> for AvailableExpressions<T> {
    const BACKWARD: bool = false;

    type Data = BTreeSet<lir::Expression>;

    fn new() -> Self {
        unreachable!()
    }

    fn new_with_metadata(cfg: &crate::cfg::Cfg<lir::Function<T>>) -> Self {
        Self {
            anticipated: analyze(cfg),
        }
    }

    fn default(&self) -> Self::Data {
        unreachable!()
    }

    fn default_with_metadata(&self, label: &Label) -> Self::Data {
        self.anticipated.outputs[label].clone()
    }

    fn transfer(&self, statement: &lir::Statement<T>, output: &mut Self::Data) {
        match statement {
            lir::Statement::Jump(_)
            | lir::Statement::CJump { .. }
            | lir::Statement::Label(_)
            | lir::Statement::Return(_) => (),
            lir::Statement::Call(_, _, returns) => {
                for r#return in 0..*returns {
                    AnticipatedExpressions::remove(output, &lir::Expression::Return(r#return));
                }
            }
            lir::Statement::Move {
                destination,
                source: _,
            } => {
                AnticipatedExpressions::remove(output, destination);
            }
        }
    }

    fn merge<'a, I>(&self, outputs: I, input: &mut Self::Data)
    where
        I: Iterator<Item = Option<&'a Self::Data>>,
        Self::Data: 'a,
    {
        <AnticipatedExpressions as Analysis<lir::Function<T>>>::merge(
            &self.anticipated.analysis,
            outputs,
            input,
        );
    }
}
