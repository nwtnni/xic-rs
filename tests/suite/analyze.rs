use xic::api::analyze::analyze;
use xic::api::analyze::display;
use xic::api::analyze::LiveRanges;
use xic::api::analyze::LiveVariables;
use xic::asm;
use xic::data::asm::Function;
use xic::data::operand::Label;
use xic::data::operand::Scale::_8;
use xic::data::operand::Temporary;
use xic::data::symbol;

macro_rules! assembly {
    ($function:ident: $arguments:tt -> $returns:tt; $($tt:tt)*) => {
        {
            let enter = Label::Fixed(symbol::intern_static("enter"));
            let exit = Label::Fixed(symbol::intern_static("exit"));

            let mut statements = assembly!($($tt)*);
            statements.insert(0, asm!((enter:)));
            statements.push(asm!((exit:)));
            statements.push(asm!((ret<$returns>)));

            Function::<Temporary> {
                name: symbol::intern_static(stringify!($function)),
                statements,
                arguments: $arguments,
                returns: $returns,
                enter,
                exit,
            }
        }
    };
    (temporaries: $($temporary:ident),+ $(,)? ; $($tt:tt)*) => {
        {
            $(let $temporary = Temporary::Fixed(symbol::intern_static(stringify!($temporary)));)*
            {
                assembly!($($tt)*)
            }
        }
    };
    (labels: $($label:ident),+ $(,)? ; $($tt:tt)*) => {
        {
            $(let $label = Label::Fixed(symbol::intern_static(stringify!($label)));)*
            {
                assembly!($($tt)*)
            }
        }
    };
    ($($statement:tt)*) => {
        vec![$(asm!($statement)),*]
    };
}

macro_rules! live_variables {
    ($function:ident $($tt:tt)*) => {
        #[test]
        fn $function() {
            let function = assembly!($function $($tt)*);
            let (live_variables, live_ranges) = live(function);
            insta::assert_display_snapshot!(live_variables);
            insta::assert_display_snapshot!(live_ranges);
        }
    }
}

fn live(function: Function<Temporary>) -> (String, String) {
    let cfg = xic::api::construct_cfg(function);
    let live_variables = analyze::<LiveVariables<_>, _>(&cfg);
    let annotated_cfg = super::graph(display(&live_variables, &cfg));
    let annotated_assembly = LiveRanges::new(&live_variables, cfg).to_string();
    (annotated_cfg, annotated_assembly)
}

live_variables! {
    smoke: 0 -> 0;
}

live_variables! {
    smoke_move: 0 -> 1;
    temporaries: x;
    (mov rax, x)
}

live_variables! {
    pass_one_receieve_one: 0 -> 1;
    labels: black_box;
    temporaries: x;
    (mov x, 5)
    (mov rdi, x)
    (call<1, 1> black_box)
    (mov x, rax)
    (mov rax, x)
}

live_variables! {
    clobber_across_call: 0 -> 2;
    labels: clobber;
    temporaries: x, y;
    (mov x, 0)
    (mov y, 1)
    (call<0, 0> clobber)
    (mov rax, x)
    (mov rdx, y)
}

live_variables! {
    clobber_across_mul: 0 -> 2;
    temporaries: x, y;
    (mov x, 1)
    (mov y, 2)
    (mov rax, x)
    (imul 5)
    (mov rdx, y)
}

live_variables! {
    clobber_across_div: 0 -> 2;
    temporaries: x, y;
    (mov x, 1)
    (mov y, 2)
    (mov rax, x)
    (cqo)
    (idiv 5)
    (mov rdx, y)
}

live_variables! {
    clobber_across_hul: 0 -> 2;
    temporaries: x, y;
    (mov x, 1)
    (mov y, 2)
    (mov rax, x)
    (ihul 5)
    (mov rax, y)
}

live_variables! {
    clobber_across_mod: 0 -> 2;
    temporaries: x, y;
    (mov x, 1)
    (mov y, 2)
    (mov rax, x)
    (cqo)
    (imod 5)
    (mov rax, y)
}

live_variables! {
    clobber_across_mul_hul: 0 -> 2;
    temporaries: x, y;
    (mov x, 1)
    (mov y, 2)
    (mov rax, x)
    (imul 3)
    (ihul 4)
    (mov rax, y)
}

live_variables! {
    clobber_across_div_mod: 0 -> 2;
    temporaries: x, y;
    (mov x, 1)
    (mov y, 2)
    (mov rax, x)
    (cqo)
    (idiv 1)
    (cqo)
    (imod 10)
    (mov rax, y)
}

live_variables! {
    redundant_moves: 0 -> 1;
    temporaries: useless;
    (mov useless, 0)
    (mov useless, 1)
    (mov useless, 2)
    (mov useless, 3)
    (mov rax, useless)
}

live_variables! {
    propagate_liveness_move: 0 -> 1;
    temporaries: x, y, z;
    (mov x, 0)
    (mov y, x)
    (mov z, y)
    (mov rdx, z)
    (mov rax, rdx)
}

live_variables! {
    propagate_liveness_neg: 0 -> 1;
    temporaries: x;
    (mov x, 1)
    (neg x)
    (neg x)
    (neg x)
    (mov rax, x)
}

live_variables! {
    propagate_liveness_memory: 0 -> 1;
    temporaries: x;
    (mov x, 1)
    (mov x, [x])
    (mov rax, x)
}

live_variables! {
    propagate_liveness_memory_two: 0 -> 1;
    temporaries: x, y;
    (mov x, 1)
    (mov y, 2)
    (mov x, [x + y * _8])
    (mov rax, x)
}

live_variables! {
    propagate_liveness_memory_tree: 0 -> 1;
    temporaries: a, b, c, d, l, r;
    (mov a, 0)
    (mov b, 1)
    (mov c, 2)
    (mov d, 3)
    (mov a, [a + 8])
    (mov b, [b + c])
    (mov c, [c * _8 + 8])
    (mov d, [d])
    (mov l, [c + d + 8])
    (mov r, [a + b * _8])
    (mov rax, [l + r * _8 + 16])
}

live_variables! {
    everything_is_meaningless: 0 -> 0;
    temporaries: x, y, z;
    (mov x, 1)
    (mov y, 2)
    (add y, x)
    (mov z, 5)
    (sub x, y)
    (mov rax, [x + z])
    (imul 5)
    (ihul 2)
    (neg rax)
    (nop)
    (mov z, rdx)
}

live_variables! {
    everything_is_meaningless_except_div_mod: 0 -> 0;
    temporaries: x, y;
    (mov x, 1)
    (mov y, 2)
    (add y, x)
    (mov rax, y)
    (cqo)
    (idiv 2)
    (mov x, rax)
    (add x, 1)
    (add x, y)
}

live_variables! {
    everything_is_meaningless_except_call: 0 -> 0;
    temporaries: x, y;
    labels: black_box;
    (mov x, 1)
    (mov y, 2)
    (add x, y)
    (add y, x)
    (mov rdi, x)
    (mov rsi, y)
    (call<2, 0> black_box)
    (add x, y)
    (add y, x)
    (mov y, rax)
    (imul x)
    (mov x, rax)
}

live_variables! {
    everything_is_meaningless_except_cmp: 0 -> 0;
    temporaries: x, y;
    labels: exit, fallthrough;
    (mov x, 1)
    (mov y, 2)
    (cmp x, y)
    (add x, 1)
    (je exit)
    (fallthrough:)
    (add y, 1)
    (jmp exit)
}

live_variables! {
    everything_is_meaningless_except_memory_write: 0 -> 0;
    temporaries: x, y;
    (mov x, 1)
    (mov y, 2)
    (mov [x + y], 5)
    (add x, 1)
    (add y, 1)
}

live_variables! {
    div_mod_cant_save_you_all_from_death: 0 -> 0;
    temporaries: x, y, useless;
    (mov useless, 0)
    (mov x, 1)
    (mov y, 2)
    (add y, x)
    (mov rax, y)
    (add useless, 5)
    (cqo)
    (idiv 2)
    (mov x, rax)
    (add x, 1)
    (add x, y)
    (add useless, 5)
}

live_variables! {
    call_cant_save_you_all_from_death: 0 -> 0;
    temporaries: x, y, useless;
    labels: black_box;
    (mov useless, 0)
    (mov x, 1)
    (mov y, 2)
    (add x, y)
    (add y, x)
    (add useless, 5)
    (mov rdi, x)
    (mov rsi, y)
    (call<2, 0> black_box)
    (add x, y)
    (add y, x)
    (mov y, rax)
    (imul x)
    (mov x, rax)
    (neg useless)
}

live_variables! {
    cmp_cant_save_you_all_from_death: 0 -> 0;
    temporaries: x, y, useless;
    labels: exit, fallthrough;
    (mov useless, 0)
    (mov x, 1)
    (add useless, 5)
    (mov y, 2)
    (cmp x, y)
    (add x, 1)
    (and useless, 1)
    (je exit)
    (fallthrough:)
    (or useless, 1)
    (add y, 1)
    (jmp exit)
}

live_variables! {
    memory_write_cant_save_you_all_from_death: 0 -> 0;
    temporaries: x, y, useless;
    (mov x, 1)
    (mov y, 2)
    (mov useless, 0)
    (add useless, x)
    (mov [x + y], 5)
    (add x, 1)
    (add y, 1)
    (add useless, y)
}
