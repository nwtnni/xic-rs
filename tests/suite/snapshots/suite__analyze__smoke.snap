---
source: tests/suite/analyze.rs
assertion_line: 99
expression: live_variables
---

   smoke

+----------+
|  enter:  |
| {rsp}    |
| jmp exit |
| {rsp}    |
+----------+
  |
  |
  v
+----------+
|  exit:   |
| {rsp}    |
| ret      |
| {}       |
+----------+

